File: frontend/src/api/notifications.ts
import { client } from './client';

export interface NotificationOut {
  id: string;
  category: string;
  title: string;
  body: string;
  link_url?: string;
  biz_type?: string;
  biz_id?: string;
  meta?: Record<string, any>;
  read_at?: string | null;
  sent_via?: string[];
  created_at: string;
}

export interface NotificationSubscription {
  category: string;
  in_app_enabled: boolean;
  email_enabled: boolean;
}

export interface ListNotificationsParams {
  unread_only?: boolean;
  category?: string;
  limit?: number;
  before_id?: string;
}

export async function listNotifications(opts: ListNotificationsParams = {}): Promise<{ items: NotificationOut[]; has_more: boolean }> {
  const params = new URLSearchParams();
  if (opts.unread_only !== undefined) params.append('unread_only', String(opts.unread_only));
  if (opts.category) params.append('category', opts.category);
  if (opts.limit) params.append('limit', String(opts.limit));
  if (opts.before_id) params.append('before_id', opts.before_id);

  const qs = params.toString();
  const url = `/api/v1/notifications${qs ? `?${qs}` : ''}`;
  const res = await client.get(url);
  return res.data;
}

export async function getUnreadCount(): Promise<{ total: number; by_category: Record<string, number> }> {
  const res = await client.get('/api/v1/notifications/unread-count');
  return res.data;
}

export async function markRead(params: { ids?: string[]; all?: boolean }): Promise<{ updated: number }> {
  const res = await client.post('/api/v1/notifications/mark-read', params);
  return res.data;
}

export async function getSubscriptions(): Promise<NotificationSubscription[]> {
  const res = await client.get('/api/v1/notifications/subscriptions');
  return res.data;
}

export async function updateSubscription(category: string, payload: { in_app_enabled: boolean; email_enabled: boolean }): Promise<NotificationSubscription> {
  const res = await client.put(`/api/v1/notifications/subscriptions/${category}`, payload);
  return res.data;
}

File: backend/app/api/v1/notifications.py
from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import CurrentUser
from app.db import get_db
from app.models import Notification, NotificationCategory
from app.schemas import (
    MarkReadRequest,
    NotificationListResponse,
    NotificationOut,
    SubscriptionOut,
    SubscriptionUpdate,
    UnreadCountResponse,
)
from app.services import notifications as svc

router = APIRouter(prefix="/notifications")


@router.get("", response_model=NotificationListResponse, tags=["notifications"])
async def list_notifications(
    user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
    unread_only: bool = False,
    category: NotificationCategory | None = None,
    limit: int = 20,
    before_id: UUID | None = None,
):
    before_created_at = None
    if before_id is not None:
        cursor = await db.get(Notification, before_id)
        if cursor is None or cursor.user_id != user.id:
            raise HTTPException(404, "notification.not_found")
        before_created_at = cursor.created_at

    page_limit = max(1, min(limit, 100))
    rows = await svc.list_notifications(
        db,
        user_id=user.id,
        unread_only=unread_only,
        category=category,
        limit=page_limit + 1,
        before_created_at=before_created_at,
    )
    items = rows[:page_limit]
    return NotificationListResponse(
        items=[NotificationOut.model_validate(item) for item in items],
        has_more=len(rows) > page_limit,
    )


@router.get("/unread-count", response_model=UnreadCountResponse, tags=["notifications"])
async def unread_count(
    user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
):
    counts = await svc.count_unread(db, user_id=user.id)
    return UnreadCountResponse.model_validate(counts)


@router.post("/mark-read", tags=["notifications"])
async def mark_read(
    payload: MarkReadRequest,
    user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
):
    updated = await svc.mark_read(
        db,
        user_id=user.id,
        notification_ids=payload.ids,
        all=payload.all,
    )
    return {"updated": updated}


@router.get("/subscriptions", response_model=list[SubscriptionOut], tags=["notifications"])
async def get_subscriptions(
    user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
):
    subscriptions = await svc.get_subscriptions(db, user_id=user.id)
    return [SubscriptionOut.model_validate(item) for item in subscriptions]


@router.put(
    "/subscriptions/{category}",
    response_model=SubscriptionOut,
    tags=["notifications"],
)
async def update_subscription(
    category: NotificationCategory,
    payload: SubscriptionUpdate,
    user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
):
    subscription = await svc.upsert_subscription(
        db,
        user_id=user.id,
        category=category,
        in_app_enabled=payload.in_app_enabled,
        email_enabled=payload.email_enabled,
    )
    return SubscriptionOut.model_validate(subscription)
